### ISSUE TITLE
Disk usage calculations

### ISSUE THREAD

**Issue by arriterx** [webmin/webmin#2084]
Hello, why are these disk usage calculations are so wrong?

![immagine](https://github.com/webmin/webmin/assets/56051435/506e5901-302e-4eb9-86b2-b9515c75cebd)

Also local disk space calculations in the image below are wrong according to the image above...
![immagine](https://github.com/webmin/webmin/assets/56051435/2dba7e13-f7c8-4ca0-8cdc-99e1001a9aec)

Can someone explain it or is that a bug of this version?

**Comment by iliajie** [comment]
Hello,

What is the output of the following command on your system:

```
df -h
```

**Comment by arriterx** [comment]
Here it is
![immagine](https://github.com/webmin/webmin/assets/56051435/744eb8da-97e2-46fe-ba33-8f2d365764e7)

**Comment by arriterx** [comment]
So, as far as I understand the inconsistensies above are not the Webmin issue, but in general an Ubuntu issues/bugs, since the screenshot below has some unexplainable for me rounding issues..., right?

**Comment by iliajie** [comment]
I don't think anything is incorrect.

> since the screenshot below has some unexplainable for me rounding issues

Try `--block-size=TB` vs `--block-size=TiB` and `--block-size=MB` vs `--block-size=MiB`, e.g.:

```
df -h --block-size=TiB
```

https://simple.wikipedia.org/wiki/Mebibyte#:~:text=A%20mebibyte%20(a%20contraction%20of,computer%20data%20storage%2C%20symbol%20MiB.&text=The%20mebibyte%20is%20closely%20related,see%20binary%20prefix%20for%20background).

**Comment by arriterx** [comment]
> Do you consider system reserver (usually 5%) space?

Yes, I have considered that, and no problem with that, since I assume totals already already include the system reserved space (so instead of total lets say 17TB i see 16.23TB) and it's completely ok. 
What is not okay is that FREE + USED != TOTAL... so on 16.23TB raid5 i loose somehow almost 1TB of space which should be there, but somehow is not there.

**Comment by iliajie** [comment]
The second screenshot provided indeed has a discrepancy in used space, which should be around 8.4 TiB and not 9.18 TiB.

I wonder now if it perhaps shows cached data? Would you mind clicking the arrow button in the top-right corner of the dashboard to force a reload of the system information? This should reflect the most up-to-date data.

**Comment by arriterx** [comment]
Now it seems you got my point.
Here are the screenshots
![immagine](https://github.com/webmin/webmin/assets/56051435/065e8f7a-019d-4826-b4db-998e28a07edd)
![immagine](https://github.com/webmin/webmin/assets/56051435/d912aebd-114b-4700-9b72-bed7058264f1)
Still all used + all free space is not equal to all total  space...

**Comment by jcameron** [comment]
Depending on the filesystem, some fraction of the used space will be reserved for the `root` user (up to 5%)

**Comment by arriterx** [comment]
> Depending on the filesystem, some fraction of the used space will be reserved for the `root` user (up to 5%)

And this part of space is not declared in any place in OS and in Webmin, right?

**Comment by jcameron** [comment]
No, we don't break it out separately. But you can see it on the Disk and Network Filesystems page ... it's the different between Free+Used and Total

**Comment by p-knulp** [comment]
Hi everybody, in dashboard when calculate the total space on disks, not include an external usb disk with NTFS partition. But it's consider only ext(n) filesystem.
Is possible plan a patch for this issue ?
Tnx!

**Comment by chris001** [comment]
Is the external USB NTFS mounted?  Can you change into its directories from the command line?

**Comment by p-knulp** [comment]
Hi Chris , yes the external USB DISK NTFS is mounted  and i can change , read, create and delete directories or files as root or other user. But in dashboard into "Disk Usage" section not view a NTFS USB mounted. So it does not calculate the total available space but only the one with ext(n) type partitions.
Tnx!

**Comment by chris001** [comment]
on your system, at the terminal, what's the output of `du -h` ?

**Comment by p-knulp** [comment]
> on your system, at the terminal, what's the output of `df -h` , does it list the `/mnt/usb-disk-ntfs` ?

Hi Chris, yes the path: "/mnt/usb-disk-ntfs"  it's in list output  of command : df -h
Tnx!

[Screenshot]
![image](https://github.com/user-attachments/assets/ce207c58-3fc0-4605-97c2-a8691cca80e6)

**Comment by chris001** [comment]
OK it's listed as `/dev/sdd1`, this might be a bug for it to not be included in the Dashboard Disk Usage, unless the reason would be because it's a USB drive so it might be categorized as a Network filesystem, not a System filesystem?

**Comment by p-knulp** [comment]
> OK it's listed as `/dev/sdd1`, this might be a bug for it to not be included in the Dashboard Disk Usage, unless the reason would be because it's a USB drive so it might be categorized as a Network filesystem, not a System filesystem?

Hi Chris, i think this is a bug of Dashboard section, (Disk Usage) and  when calculate the total space disks attached at machine USB and not USB - ref.: Dashboard section, (System Info -> Local disk space ) -  because "/dev/sdb1" , "/dev/sdc1" they are ext4 type partition an not NTFS. 
Anyway, if umount and unplugged every USB Disk type partition ext4 and NTFS ,  and after plugged and mount only USB NTFS type partition, this last will be "/dev/sdb1/" and always NOT be included in the Dashboard  section System Info -> Local disk space  and Disk Usage, because maybe a NTFS type partition. (P.S.: same problem if switch a power off and power on  of the physical machine. Every USB Disk attached ext4 or NTFS they are categorized as same type filesystem.
Tnx!

**Comment by iliajie** [comment]
@p-knulp What is the output of `cat /etc/mtab` on your system?

**Comment by iliajie** [comment]
@p-knulp Also, if you edit `mount-lib.pl` file under `mount` Webmin installation directory, find line:

```
$m->[2] =~ /^ext/ ||
```

.. replace it with:

```
$m->[2] =~ /^ext/ || $m->[2] =~ /^ntfs/ ||
```

.. restart Webmin manually and force-refresh system information page on the dashboard — does it just work _right_ then for you?

**Comment by p-knulp** [comment]
> @p-knulp What is the output of `cat /etc/mtab` on your system?

Hi iliajie,  this is the output, in this time

![image](https://github.com/user-attachments/assets/675f59ed-ec58-49a3-8bc0-83eff82eb03f)

**Comment by p-knulp** [comment]
> @p-knulp Also, if you edit `mount-lib.pl` file under `mount` Webmin installation directory, find line:
> 
> ```
> $m->[2] =~ /^ext/ ||
> ```
> 
> .. replace it with:
> 
> ```
> $m->[2] =~ /^ext/ || $m->[2] =~ /^ntfs/ ||
> ```
> 
> .. restart Webmin manually and force-refresh system information page on the dashboard — does it just work _right_ then for you?

Hi, now I'll try this modification you suggested

**Comment by iliajie** [comment]
Based on the output, it seems that the filesystem should be `vfat` rather than `ntfs`.

**Comment by iliajie** [comment]
...or possibly even `fuseblk`, though I'm not certain if that would work right out of the box.

**Comment by p-knulp** [comment]
> Based on the output, it seems that the filesystem should be `vfat` rather than `ntfs`.

I know the partitions of "/dev/sdb1" it's NTFS. For debian read and write on any partition like windows with fuseblk.

But  i can see /dev/sdb1 it's a NTFS when i go in Disk and Network Filesystems   

![image](https://github.com/user-attachments/assets/f5ee4a34-2366-4467-beac-9ffce097c97d)

**Comment by iliajie** [comment]
Try `fuseblk` ..

**Comment by p-knulp** [comment]
> .. or possibly even `fuseblk`, though I'm not certain if that would work right out of the box.


> Try `fuseblk` ..

Yes I also think that the modification is with "fuseblk" instead of "ntfs".
Tnx!

**Comment by iliajie** [comment]
> Yes I also think that the modification is with "fuseblk" instead of "ntfs".

Will you be able to provide a feedback? Could you also check top command to see how that goes?

**Comment by p-knulp** [comment]
I have edit perl file "mount-lib.pl" add  a ||  condition like this : 

(about  to row number 320...)
**# Add up all local filesystems
foreach my $m (@mounted) {
	if ($m->[2] =~ /^ext/ || $m->[2] eq "fuseblk" ||**
	
Restart Webmin and now it's ok !

![image](https://github.com/user-attachments/assets/e75c4c62-2814-4169-a377-56ce8d39ddc8)

Tnx!

**Comment by p-knulp** [comment]
> > Yes I also think that the modification is with "fuseblk" instead of "ntfs".
> 
> Will you be able to provide a feedback? Could you also check `top` command output to see how that goes?

Yes it's work now.

output top command ? :


![image](https://github.com/user-attachments/assets/1e78e52c-3f8d-4837-a9dc-91628c155c29)

**Comment by chris001** [comment]
Should it be easy to add ZFS, BTRFS, XFS, etc?

**Comment by p-knulp** [comment]
> Should it be easy to add ZFS, BTRFS, XFS, etc?

Hi Chris, I definitely think so. Another type of partition format is "ExtFAT". In fact, with webmin it doesn't seem possible to create a mount point of type ExtFAT and obviously mount it if not from commandline.

**Comment by iliajie** [comment]
> > Should it be easy to add ZFS, BTRFS, XFS, etc?
> 
> Hi Chris, I definitely think so.

Don't we already support these filesystems?

**Comment by p-knulp** [comment]
> > > Should it be easy to add ZFS, BTRFS, XFS, etc?
> > 
> > 
> > Hi Chris, I definitely think so.
> 
> Don't we already support these filesystems?

Hi Chris and Iliajie,
looking a part of source code : '..mnt/mount-lib.pl', seem that already support some these filesystems!

![image](https://github.com/user-attachments/assets/e28592d2-7941-4f45-a161-7cb28dcdff7d)

**Comment by p-knulp** [comment]
> > > Yes I also think that the modification is with "fuseblk" instead of "ntfs".
> > 
> > 
> > Will you be able to provide a feedback? Could you also check `top` command output to see how that goes?
> 
> Yes i can also check `top` command output to see how that goes! Tnx!

Hi Iliajie, I'm monitoring the output of the "top" command and so far I don't see any abnormal situations, after modifying the source code of mount-lib.pl.

**Comment by chris001** [comment]
> Don't we already support these filesystems?

Yes!  However, the MacOS filesystems are missing.  Would this help? [How to mount APFS on Linux](https://superuser.com/questions/1267291/how-to-mount-apfs-on-linux-or-windows/).

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
